### PR TITLE
[CI] Support reproduceable build for git clone (#6562)

### PR DESCRIPTION
**Pull request by liushilongbuaa**
* support reproduceable build for git clone

Signed-off-by: shilongliu <shilongliu@microsoft.com>

#### Why I did it
cherry-pick reproducible build for git to 202012.
#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012

### CODE REVIEW

File: src/sonic-build-hooks/hooks/git [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    result=$?
+
+    # if sub command is not "clone", exit
+    if [[ $MODE_CLONE != 1 ]];then
+        exit $result
+    fi
+
+    get_clone_path "$@"
+    pushd $clone_PATH &> /dev/null
+    commit_latest=`$REAL_COMMAND log -n 1 | head -n 1| awk '{print$2}'`
```
Comment (xumia) [inline src/sonic-build-hooks/hooks/git:59]: How about using "git rev-parse HEAD"?

File: src/sonic-build-hooks/hooks/git [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    [ -f $version_file ] && commit=`grep $URL $version_file | awk -F, '{print$2}'`
+
+    # control version or record version file
+    if [[ $ENABLE_VERSION_CONTROL_GIT == "y" ]];then
+        # control version
+        [ ! -z $commit ] && $REAL_COMMAND reset --hard $commit &> /dev/null
+    else
+        # record version file
+        echo "$URL==$commit_latest" >> $new_version_file
```
Comment (xumia) [inline src/sonic-build-hooks/hooks/git:68]: Maybe We can record version whether ENABLE_VERSION_CONTROL_GIT enabled or not, right?

File: src/sonic-build-hooks/hooks/git [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    # if sub command is not "clone", exit
+    if [[ $MODE_CLONE != 1 ]];then
+        exit $result
+    fi
+
+    get_clone_path "$@"
+    pushd $clone_PATH &> /dev/null
+    commit_latest=`$REAL_COMMAND log -n 1 | head -n 1| awk '{print$2}'`
+    [ -f $version_file ] && commit=`grep $URL $version_file | awk -F, '{print$2}'`
```
Comment (xumia) [inline src/sonic-build-hooks/hooks/git:60]: Do we need to do something when the version does not exist?

File: src/sonic-build-hooks/hooks/git [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+parse_config(){
+    . /usr/local/share/buildinfo/scripts/buildinfo_base.sh
+    REAL_COMMAND=$(get_command git)
+
+    version_file=$VERSION_PATH/versions-git
+    new_version_file=$BUILD_VERSION_PATH/versions-git
+
+    MODE_CLONE=0
+    # parse input parameters
+    for i in "$@"
```
Comment (xumia) [inline src/sonic-build-hooks/hooks/git:12]: How about using a meaningful value? like param.

File: src/sonic-build-hooks/hooks/git [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    while (( "$#" )); do
+        case $1 in
+            -b|--branch|--reference|--reference-if-able|-c|--config|--origin|-u|--upload-pack|-j|--jobs|--depth|--dissociate)
+                shift 2
+                ;;
+            clone|-l|--local|--no-hardlinks|-s|--shared|--dissociate|-q|--quiet|-v|--verbose|--progress|--server-option=*|--bare|--sparse|--filter=*|--template=*|--mirror|--reference|--shallow-*|--no-tags|--recurse-submodules*|--remote-submodules|--no-remote-submodules|--separate-git-dir*)
+                shift 1
+                ;;
+            *)
+                if [[ $URL == "" ]];then
```
Comment (xumia) [inline src/sonic-build-hooks/hooks/git:31]: Looks like you try to exclude all the known parameters, then do an assumption, the value is URL, is it true?
How about using the web pattern? like git: http: https, something like it.
How about excluding all the parameter starts with "-"

File: src/sonic-build-hooks/hooks/git [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+                else
+                    clone_PATH=$1
+                fi
+                shift 1
+                ;;
+        esac
+    done
+
+    # if not specific clone path, get default clone path
+    [ -z $clone_PATH ] && clone_PATH=`echo $URL | rev | awk -F/ '{print$1}' | rev | awk -F. '{print$1}'`
```
Comment (xumia) [inline src/sonic-build-hooks/hooks/git:42]: The command awk supports to get the last field, $NF. How about to change as below? Or any better regex by sed?
```
[ -z $clone_PATH ] && clone_PATH=`echo $URL | awk -F/ '{print$NF}' | awk -F. '{print$1}'`
```

File: src/sonic-build-hooks/hooks/git (hunk 5 above)
Comment (liushilongbuaa) [reply src/sonic-build-hooks/hooks/git:31]: no. we need to distinguish 2 scenario:
1. --depth 2
2. --quiet

**Comment by liushilongbuaa** [comment]
@xumia comment placed in another PR:
https://github.com/Azure/sonic-buildimage/pull/7977

**Pull request closed without merging**